### PR TITLE
Simplify and streamline GetWorkflowLease logic

### PR DESCRIPTION
**Pull request by ychebotarev**
## What changed?

Remove "consistency check" parameter from GetAndUpdateWorkflow and GetWorkflowLease
To support those cases where is was still used - old version is preserved with different function name.
Also remove dead code and dead tests.

## Why?

Current implementation has some issues.
a. It is hard to understand, which can (and does) lead to hard-to-detect bugs

b. Predicate (I'm not calling it "consistency check" because it is not) is used only in just 2 cases out of > 70. In all other cases “do nothing” check is provided.

c. Current implementation of predicate check introduce huge “side effect” (clear WF context) which (most likely) has little to do with loading mutable state. I'm still not sure it should be used in the way it is used.

d. It is hard to tell if current implementation do anything useful - it loads mutable state, and if predicate failed - clear context and load it again. That is all. It is not clear that clear WF context somehow affect mutable state loading at all…

e. Logic in one (or two) consistency check function is complicated and clearly out of boundaries for "consistency check"scope. It looks more like business logic. Thus introduce more side-effects.

## How did you test it?

unit test. add a bit more tests to cover clock related logic.

## Potential risks

This is core logic, and has side-effects. I try to minimize the risk by keeping the old version intact, but it still exist.

## Is hotfix candidate?

No

### CODE REVIEW

File: service/history/api/consistency_checker.go [hunk 1]
```diff
@@ -56,6 +55,13 @@ type (
 			lockPriority workflow.LockPriority,
 		) (string, error)
 		GetWorkflowLease(
+			ctx context.Context,
+			reqClock *clockspb.VectorClock,
+			workflowKey definition.WorkflowKey,
+			lockPriority workflow.LockPriority,
+		) (WorkflowLease, error)
+
+		GetWorkflowLeaseWithConsistancyCheck(
```
Comment (yycptt) [inline service/history/api/consistency_checker.go:64]:
```suggestion
		GetWorkflowLeaseWithConsistencyCheck(
```

File: service/history/api/consistency_checker.go [hunk 2]
```diff
@@ -103,62 +109,57 @@ func (c *WorkflowConsistencyCheckerImpl) GetCurrentRunID(
...
-	workflowKey definition.WorkflowKey,
-	lockPriority workflow.LockPriority,
-) (WorkflowLease, error) {
+) error {
+	if reqClock == nil {
+		return nil
+	}
+	currentClock := c.shardContext.CurrentVectorClock()
+	if !vclock.Comparable(reqClock, currentClock) {
+		// request vector clock cannot is not comparable with current shard vector clock
```
Comment (yycptt) [inline service/history/api/consistency_checker.go:139]:
```suggestion
		// request vector clock is not comparable with current shard vector clock
```

File: service/history/api/update_workflow_util.go [hunk 3]
```diff
@@ -60,6 +58,31 @@ func GetAndUpdateWorkflowWithNew(
 	return UpdateWorkflowWithNew(shard, ctx, workflowLease, action, newWorkflowFn)
 }
 
+func GetAndUpdateWorkflowWithConsistancyCheck(
```
Comment (yycptt) [inline service/history/api/update_workflow_util.go:61]:
```suggestion
func GetAndUpdateWorkflowWithConsistencyCheck(
```

File: service/history/api/consistency_checker.go [hunk 4]
```diff
@@ -178,9 +179,91 @@ func (c *WorkflowConsistencyCheckerImpl) getWorkflowLeaseValidatedByClock(
...
+	if currentRunID == workflowLease.GetContext().GetWorkflowKey().RunID {
+		return workflowLease, nil
+	}
+
+	workflowLease.GetReleaseFn()(nil)
+	return nil, consts.ErrLocateCurrentWorkflowExecution
+}
+
+// The code below is an old implementation of GetWorkflowLease
+// It should be avoided. We should revisit the logic
```
Comment (yycptt) [inline service/history/api/consistency_checker.go:226]: Plz update the comment. Consistency is necessary when shard clock is not applicable and request can't operate on stale mutable state.

File: service/history/api/consistency_checker.go [hunk 5]
```diff
@@ -178,9 +179,91 @@ func (c *WorkflowConsistencyCheckerImpl) getWorkflowLeaseValidatedByClock(
...
+			return c.getWorkflowLeaseValidatedByClock(
+				ctx,
+				reqClock,
+				currentClock,
+				workflowKey,
+				lockPriority,
+			)
+		}
+		// request vector clock cannot is not comparable with current shard vector clock
+		// use methods below
```
Comment (yycptt) [inline service/history/api/consistency_checker.go:247]: Can we have only one copy of logic for `reqClock != nil` case?

File: service/history/api/consistency_checker.go [hunk 6]
```diff
@@ -178,9 +179,91 @@ func (c *WorkflowConsistencyCheckerImpl) getWorkflowLeaseValidatedByClock(
 		release(err)
 		return nil, err
 	}
+
 	return NewWorkflowLease(wfContext, release, mutableState), nil
 }
 
+func (c *WorkflowConsistencyCheckerImpl) getCurrentWorkflowContext(
```
Comment (yycptt) [inline service/history/api/consistency_checker.go:186]: Can this reuse the impl in getCurrentWorkflowContextWithConsistencyCheck? They are almost identical.

Comment (ychebotarev) [reply service/history/api/consistency_checker.go:182]: in the first version idea was to keep the old code. 
I updated the PR, and try to reuse it.

File: service/history/api/consistency_checker.go (hunk 5 above)
Comment (ychebotarev) [reply service/history/api/consistency_checker.go]: in the first iteration the idea was to keep old code intact.
Yes, if we fine - I will combine the code.

File: service/history/api/consistency_checker.go (hunk 4 above)
Comment (ychebotarev) [reply service/history/api/consistency_checker.go]: done

File: service/history/api/consistency_checker.go [hunk 7]
```diff
@@ -101,97 +106,129 @@ func (c *WorkflowConsistencyCheckerImpl) GetCurrentRunID(
...
-		return c.getWorkflowLeaseValidatedByCheck(
-			ctx,
-			consistencyPredicate,
-			workflowKey,
-			lockPriority,
-		)
+		return c.getWorkflowLease(ctx, consistencyPredicate, workflowKey, lockPriority)
 	}
+
 	return c.getCurrentWorkflowContext(
```
Comment (yycptt) [inline service/history/api/consistency_checker.go:146]: nit: can you help rename this to `getCurrentWorkflowLease`? I guess it was missed in the previous rename PR (workflowContext -> workflowLease)

File: service/history/api/consistency_checker.go [hunk 8]
```diff
@@ -226,80 +264,44 @@ func (c *WorkflowConsistencyCheckerImpl) getWorkflowLeaseValidatedByCheck(
 	return NewWorkflowLease(wfContext, release, mutableState), nil
 }
 
-func (c *WorkflowConsistencyCheckerImpl) getCurrentWorkflowContext(
+func (c *WorkflowConsistencyCheckerImpl) getWorkflowLeaseValidatedByClock(
```
Comment (yycptt) [inline service/history/api/consistency_checker.go:267]: looks like this method is no longer being used and can be removed.

File: service/history/api/consistency_checker.go [hunk 9]
```diff
@@ -101,97 +106,129 @@ func (c *WorkflowConsistencyCheckerImpl) GetCurrentRunID(
...
-			workflowKey,
-			lockPriority,
-		)
+		return c.getWorkflowLease(ctx, consistencyPredicate, workflowKey, lockPriority)
 	}
+
 	return c.getCurrentWorkflowContext(
 		ctx,
-		consistencyPredicate,
+		nil,
```
Comment (yycptt) [inline service/history/api/consistency_checker.go:148]: We should pass in `consistencyPredicate` here I think.

File: service/history/api/consistency_checker.go (hunk 8 above)
Comment (ychebotarev) [reply service/history/api/consistency_checker.go:267]: removed

File: service/history/api/consistency_checker.go (hunk 9 above)
Comment (ychebotarev) [reply service/history/api/consistency_checker.go:148]: good catch

File: service/history/api/consistency_checker.go (hunk 7 above)
Comment (ychebotarev) [reply service/history/api/consistency_checker.go:146]: sure, renamed

**Review by yycptt — APPROVED** [review]

**Pull request merged**